File: apps/js/src/lib/sqid.ts
import Sqids from "sqids";

import { slugify } from "#/lib/slug";

const MIN_LENGTH = 6;
const MIN_ALPHABET = 16;

let shared: Sqids | undefined;

function sqids(): Sqids {
	if (shared === undefined) {
		const alphabet = process.env["SQIDS_ALPHABET"] ?? "";

		if (alphabet.length < MIN_ALPHABET) {
			throw new RangeError("sqid alphabet must be at least 16 characters");
		}

		shared = new Sqids({ alphabet, minLength: MIN_LENGTH });
	}

	return shared;
}

export function encode(id: number): string {
	if (!Number.isSafeInteger(id) || id < 0) {
		throw new RangeError("sqid ids must be non-negative safe integers");
	}

	return sqids().encode([id]);
}

export function decode(s: string): number | undefined {
	const ids = sqids().decode(s);
	const [id] = ids;

	if (ids.length !== 1 || id === undefined || !Number.isSafeInteger(id)) {
		return undefined;
	}

	return sqids().encode([id]) === s ? id : undefined;
}

export function productPath(sqid: string, name: string): string {
	return `/products/${sqid}/${slugify(name)}`;
}

File: apps/go/internal/sqid/sqid.go
package sqid

import (
	"errors"
	"math"

	sqids "github.com/sqids/sqids-go"
)

const (
	minLength   = 6
	minAlphabet = 16
)

var (
	ErrInvalid  = errors.New("invalid sqid")
	ErrAlphabet = errors.New("sqid alphabet must be at least 16 characters")
)

type Codec struct {
	sqids *sqids.Sqids
}

func New(alphabet string) (*Codec, error) {
	if len(alphabet) < minAlphabet {
		return nil, ErrAlphabet
	}

	s, err := sqids.New(sqids.Options{Alphabet: alphabet, MinLength: minLength})
	if err != nil {
		return nil, err
	}

	return &Codec{sqids: s}, nil
}

func (c *Codec) Encode(id int64) (string, error) {
	if id < 0 {
		return "", ErrInvalid
	}

	return c.sqids.Encode([]uint64{uint64(id)})
}

func (c *Codec) Decode(s string) (int64, error) {
	ids := c.sqids.Decode(s)
	if len(ids) != 1 || ids[0] > math.MaxInt64 {
		return 0, ErrInvalid
	}

	canonical, err := c.sqids.Encode(ids)
	if err != nil || canonical != s {
		return 0, ErrInvalid
	}

	return int64(ids[0]), nil
}
